How to load tables to dataset in tokyo region?

I got similar error with the latest version of google-api-client(v0.20.1).
I don't think you can load data into dataset in Tokyo region for now.

> * If your data is in a location other than the US or EU multi-region, you must specify the location when you perform actions such as loading data, querying data, and exporting data.
> * When you use the API, specify your region in the location property in the jobReference section of the job resource.
https://cloud.google.com/bigquery/docs/dataset-locations#specifying_your_location

I'm not aware what "data" means, but we may need to specify `location` at all places creating `job_reference`.
I've not verified, though.
```ruby
body = {
  job_reference: {
  project_id: @project,
  job_id: job_id,
  location: 'asia-northeast1'
},
```
https://github.com/embulk/embulk-output-bigquery/blob/v0.4.6/lib/embulk/output/bigquery/bigquery_client.rb#L252

```
org.embulk.exec.PartialExecutionException: org.jruby.exceptions.RaiseException: (Error) failed to load /var/folders/tq/cbh1qmf15lz2g3ltp8pp9vdw0000gn/T/embulk_output_bigquery_20180501-73874-1tlegk3.73874.2000.csv.gz to abcdefg-777:tokyo.LOAD_TEMP_a86db79a_71ef_4651_9800_8e88b254b6e0_demotable, response:{:status_code=>404, :message=>"notFound: Not found: Job abcdefg-777:embulk_load_job_57ed72ae-6d28-4558-8848-2cd22a43616b", :error_class=>Google::Apis::ClientError}
	at org.embulk.exec.BulkLoader$LoaderState.buildPartialExecuteException(BulkLoader.java:339)
	at org.embulk.exec.BulkLoader.doRun(BulkLoader.java:565)
	at org.embulk.exec.BulkLoader.access$000(BulkLoader.java:34)
	at org.embulk.exec.BulkLoader$1.run(BulkLoader.java:352)
	at org.embulk.exec.BulkLoader$1.run(BulkLoader.java:349)
	at org.embulk.spi.Exec.doWith(Exec.java:22)
	at org.embulk.exec.BulkLoader.run(BulkLoader.java:349)
	at org.embulk.EmbulkEmbed.run(EmbulkEmbed.java:161)
	at org.embulk.EmbulkRunner.runInternal(EmbulkRunner.java:292)
	at org.embulk.EmbulkRunner.run(EmbulkRunner.java:156)
	at org.embulk.cli.EmbulkRun.runSubcommand(EmbulkRun.java:436)
	at org.embulk.cli.EmbulkRun.run(EmbulkRun.java:91)
	at org.embulk.cli.Main.main(Main.java:26)
Caused by: org.jruby.exceptions.RaiseException: (Error) failed to load /var/folders/tq/cbh1qmf15lz2g3ltp8pp9vdw0000gn/T/embulk_output_bigquery_20180501-73874-1tlegk3.73874.2000.csv.gz to abcdefg-777:tokyo.LOAD_TEMP_a86db79a_71ef_4651_9800_8e88b254b6e0_demotable, response:{:status_code=>404, :message=>"notFound: Not found: Job abcdefg-777:embulk_load_job_57ed72ae-6d28-4558-8848-2cd22a43616b", :error_class=>Google::Apis::ClientError}
	at RUBY.block in load(/Users/akama/data/embulk-output-bigquery/lib/embulk/output/bigquery/bigquery_client.rb:235)
	at RUBY.with_job_retry(/Users/akama/data/embulk-output-bigquery/lib/embulk/output/bigquery/bigquery_client.rb:56)
	at RUBY.load(/Users/akama/data/embulk-output-bigquery/lib/embulk/output/bigquery/bigquery_client.rb:167)
	at RUBY.block in load_in_parallel(/Users/akama/data/embulk-output-bigquery/lib/embulk/output/bigquery/bigquery_client.rb:155)

Error: org.jruby.exceptions.RaiseException: (Error) failed to load /var/folders/tq/cbh1qmf15lz2g3ltp8pp9vdw0000gn/T/embulk_output_bigquery_20180501-73874-1tlegk3.73874.2000.csv.gz to abcdefg-777:tokyo.LOAD_TEMP_a86db79a_71ef_4651_9800_8e88b254b6e0_demotable, response:{:status_code=>404, :message=>"notFound: Not found: Job abcdefg-777:embulk_load_job_57ed72ae-6d28-4558-8848-2cd22a43616b", :error_class=>Google::Apis::ClientError}
```

--- a/README.md
+++ b/README.md
@@ -44,6 +44,7 @@ v0.3.x has incompatibility changes with v0.2.x. Please see [CHANGELOG.md](CHANGE
 |  json_keyfile                        | string      | required when auth_method is json_key     |   | Fullpath of json key |
 |  project                             | string      | required if json_keyfile is not given     |   | project_id |
 |  dataset                             | string      | required   |                          | dataset |
+|  location                            | string      | optional   | nil                      | geographic location of dataset. See [Location](#location) |
 |  table                               | string      | required   |                          | table name, or table name with a partition decorator such as `table_name$20160929`|
 |  auto_create_dataset                 | boolean     | optional   | false                    | automatically create dataset |
 |  auto_create_table                   | boolean     | optional   | false                    | See [Dynamic Table Creating](#dynamic-table-creating) |
@@ -122,6 +123,14 @@ out:
   compression: GZIP
   source_format: NEWLINE_DELIMITED_JSON
 ```
+
+### location
+
+The geographic location of the dataset. Required except for US and EU.
+
+`auto_create_table` isn't supported except for US and EU. And GCS bucket should be in same region when you use `gcs_bucket`.
+
+See also [Dataset Locations | BigQuery | Google Cloud](https://cloud.google.com/bigquery/docs/dataset-locations)
 
 ### mode
 

--- a/lib/embulk/output/bigquery.rb
+++ b/lib/embulk/output/bigquery.rb
@@ -39,6 +39,7 @@ module Embulk
           'json_keyfile'                   => config.param('json_keyfile',                  LocalFile, :default => nil),
           'project'                        => config.param('project',                        :string,  :default => nil),
           'dataset'                        => config.param('dataset',                        :string),
+          'location'                       => config.param('location',                       :string,  :default => nil),
           'table'                          => config.param('table',                          :string),
           'dataset_old'                    => config.param('dataset_old',                    :string,  :default => nil),
           'table_old'                      => config.param('table_old',                      :string,  :default => nil),
@@ -110,6 +111,17 @@ module Embulk
           end
           task['dataset_old'] ||= task['dataset']
           task['table_old']   ||= task['table']
+        end
+
+        unless task['location'].nil?
+          task['location'] = task['location'].downcase
+          # google-api-client doesn't support create bucket with region
+          # We need to use Cloud Storage Client Libraries to support it
+          if task['auto_create_gcs_bucket']
+            unless %w[us eu].include?(task['location'])
+              raise ConfigError.new "`auto_create_gcs_bucket` isn't supported excepts in us/eu"
+            end
+          end
         end
 
         if task['table_old']

--- a/lib/embulk/output/bigquery/bigquery_client.rb
+++ b/lib/embulk/output/bigquery/bigquery_client.rb
@@ -20,6 +20,7 @@ module Embulk
           @project = @task['project']
           @dataset = @task['dataset']
           @location = @task['location']
+          @location_for_log = @location.nil? ? 'us/eu' : @location
 
           @task['source_format'] ||= 'CSV'
           @task['max_bad_records'] ||= 0
@@ -83,7 +84,7 @@ module Embulk
               else
                 job_id = "embulk_load_job_#{SecureRandom.uuid}"
               end
-              Embulk.logger.info { "embulk-output-bigquery: Load job starting... job_id:[#{job_id}] #{object_uris} => #{@project}:#{@dataset}.#{table}" }
+              Embulk.logger.info { "embulk-output-bigquery: Load job starting... job_id:[#{job_id}] #{object_uris} => #{@project}:#{@dataset}.#{table} in #{@location_for_log}" }
 
               body = {
                 job_reference: {
@@ -133,7 +134,7 @@ module Embulk
               Embulk.logger.error {
                 "embulk-output-bigquery: insert_job(#{@project}, #{body}, #{opts}), response:#{response}"
               }
-              raise Error, "failed to load #{object_uris} to #{@project}:#{@dataset}.#{table}, response:#{response}"
+              raise Error, "failed to load #{object_uris} to #{@project}:#{@dataset}.#{table} in #{@location_for_log}, response:#{response}"
             end
           end
         end
@@ -178,7 +179,7 @@ module Embulk
                 else
                   job_id = "embulk_load_job_#{SecureRandom.uuid}"
                 end
-                Embulk.logger.info { "embulk-output-bigquery: Load job starting... job_id:[#{job_id}] #{path} => #{@project}:#{@dataset}.#{table}" }
+                Embulk.logger.info { "embulk-output-bigquery: Load job starting... job_id:[#{job_id}] #{path} => #{@project}:#{@dataset}.#{table} in #{@location_for_log}" }
               else
                 Embulk.logger.info { "embulk-output-bigquery: Load job starting... #{path} does not exist, skipped" }
                 return
@@ -239,7 +240,7 @@ module Embulk
               Embulk.logger.error {
                 "embulk-output-bigquery: insert_job(#{@project}, #{body}, #{opts}), response:#{response}"
               }
-              raise Error, "failed to load #{path} to #{@project}:#{@dataset}.#{table}, response:#{response}"
+              raise Error, "failed to load #{path} to #{@project}:#{@dataset}.#{table} in #{@location_for_log}, response:#{response}"
             end
           end
         end
@@ -354,7 +355,7 @@ module Embulk
         def create_dataset(dataset = nil, reference: nil)
           dataset ||= @dataset
           begin
-            Embulk.logger.info { "embulk-output-bigquery: Create dataset... #{@project}:#{dataset}" }
+            Embulk.logger.info { "embulk-output-bigquery: Create dataset... #{@project}:#{dataset} in #{@location_for_log}" }
             hint = {}
             if reference
               response = get_dataset(reference)
@@ -370,7 +371,7 @@ module Embulk
               body[:location] = @location
             end
             opts = {}
-            Embulk.logger.debug { "embulk-output-bigquery: insert_dataset(#{@project}, #{dataset}, #{body}, #{opts})" }
+            Embulk.logger.debug { "embulk-output-bigquery: insert_dataset(#{@project}, #{dataset}, #{@location_for_log}, #{body}, #{opts})" }
             with_network_retry { client.insert_dataset(@project, body, opts) }
           rescue Google::Apis::ServerError, Google::Apis::ClientError, Google::Apis::AuthorizationError => e
             if e.status_code == 409 && /Already Exists:/ =~ e.message
@@ -382,14 +383,14 @@ module Embulk
             Embulk.logger.error {
               "embulk-output-bigquery: insert_dataset(#{@project}, #{body}, #{opts}), response:#{response}"
             }
-            raise Error, "failed to create dataset #{@project}:#{dataset}, response:#{response}"
+            raise Error, "failed to create dataset #{@project}:#{dataset} in #{@location_for_log}, response:#{response}"
           end
         end
 
         def get_dataset(dataset = nil)
           dataset ||= @dataset
           begin
-            Embulk.logger.info { "embulk-output-bigquery: Get dataset... #{@project}:#{@dataset}" }
+            Embulk.logger.info { "embulk-output-bigquery: Get dataset... #{@project}:#{dataset}" }
             with_network_retry { client.get_dataset(@project, dataset) }
           rescue Google::Apis::ServerError, Google::Apis::ClientError, Google::Apis::AuthorizationError => e
             if e.status_code == 404
@@ -432,7 +433,7 @@ module Embulk
             end
 
             opts = {}
-            Embulk.logger.debug { "embulk-output-bigquery: insert_table(#{@project}, #{dataset}, #{body}, #{opts})" }
+            Embulk.logger.debug { "embulk-output-bigquery: insert_table(#{@project}, #{dataset}, #{@location_for_log}, #{body}, #{opts})" }
             with_network_retry { client.insert_table(@project, dataset, body, opts) }
           rescue Google::Apis::ServerError, Google::Apis::ClientError, Google::Apis::AuthorizationError => e
             if e.status_code == 409 && /Already Exists:/ =~ e.message
@@ -442,9 +443,9 @@ module Embulk
 
             response = {status_code: e.status_code, message: e.message, error_class: e.class}
             Embulk.logger.error {
-              "embulk-output-bigquery: insert_table(#{@project}, #{dataset}, #{body}, #{opts}), response:#{response}"
+              "embulk-output-bigquery: insert_table(#{@project}, #{dataset}, #{@location_for_log}, #{body}, #{opts}), response:#{response}"
             }
-            raise Error, "failed to create table #{@project}:#{dataset}.#{table}, response:#{response}"
+            raise Error, "failed to create table #{@project}:#{dataset}.#{table} in #{@location_for_log}, response:#{response}"
           end
         end
 

--- a/lib/embulk/output/bigquery/bigquery_client.rb
+++ b/lib/embulk/output/bigquery/bigquery_client.rb
@@ -19,6 +19,7 @@ module Embulk
           reset_fields(fields) if fields
           @project = @task['project']
           @dataset = @task['dataset']
+          @location = @task['location']
 
           @task['source_format'] ||= 'CSV'
           @task['max_bad_records'] ||= 0
@@ -110,6 +111,10 @@ module Embulk
                   }
                 }
               }
+
+              if @location
+                body[:job_reference][:location] = @location
+              end
               
               if @task['schema_update_options']
                 body[:configuration][:load][:schema_update_options] = @task['schema_update_options']
@@ -205,6 +210,10 @@ module Embulk
                 }
               }
 
+              if @location
+                body[:job_reference][:location] = @location
+              end
+
               if @task['schema_update_options']
                 body[:configuration][:load][:schema_update_options] = @task['schema_update_options']
               end
@@ -269,6 +278,10 @@ module Embulk
                 }
               }
 
+              if @location
+                body[:job_reference][:location] = @location
+              end
+
               opts = {}
               Embulk.logger.debug { "embulk-output-bigquery: insert_job(#{@project}, #{body}, #{opts})" }
               response = with_network_retry { client.insert_job(@project, body, opts) }
@@ -312,7 +325,7 @@ module Embulk
                 "job_id:[#{job_id}] elapsed_time:#{elapsed.to_f}sec status:[#{status}]"
               }
               sleep wait_interval
-              _response = with_network_retry { client.get_job(@project, job_id) }
+              _response = with_network_retry { client.get_job(@project, job_id, location: @location) }
             end
           end
 
@@ -353,6 +366,9 @@ module Embulk
                 dataset_id: dataset,
               },
             }.merge(hint)
+            if @location
+              body[:location] = @location
+            end
             opts = {}
             Embulk.logger.debug { "embulk-output-bigquery: insert_dataset(#{@project}, #{dataset}, #{body}, #{opts})" }
             with_network_retry { client.insert_dataset(@project, body, opts) }

--- a/lib/embulk/output/bigquery/helper.rb
+++ b/lib/embulk/output/bigquery/helper.rb
@@ -67,6 +67,7 @@ module Embulk
           elements = [
             Digest::MD5.file(path).hexdigest,
             task['dataset'],
+            task['location'],
             task['table'],
             fields,
             task['source_format'],

--- a/test/test_configure.rb
+++ b/test/test_configure.rb
@@ -49,6 +49,7 @@ module Embulk
         assert_equal nil, task['json_keyfile']
         assert_equal "your_project_name", task['project']
         assert_equal "your_dataset_name", task['dataset']
+        assert_equal nil, task['location']
         assert_equal "your_table_name", task['table']
         assert_equal nil, task['dataset_old']
         assert_equal nil, task['table_old']
@@ -100,6 +101,20 @@ module Embulk
         assert_nothing_raised { Bigquery.configure(config, schema, processor_count) }
 
         config = least_config.merge('mode' => 'replace_backup')
+        assert_raise { Bigquery.configure(config, schema, processor_count) }
+      end
+
+      def test_location
+        config = least_config.merge('location' => 'us')
+        assert_nothing_raised { Bigquery.configure(config, schema, processor_count) }
+
+        config = least_config.merge('location' => 'eu')
+        assert_nothing_raised { Bigquery.configure(config, schema, processor_count) }
+
+        config = least_config.merge('location' => 'asia-northeast1')
+        assert_nothing_raised { Bigquery.configure(config, schema, processor_count) }
+
+        config = least_config.merge('location' => 'asia-northeast1', 'auto_create_gcs_bucket' => true)
         assert_raise { Bigquery.configure(config, schema, processor_count) }
       end
 

--- a/test/test_helper.rb
+++ b/test/test_helper.rb
@@ -94,6 +94,7 @@ module Embulk
       def test_create_load_job_id
         task = {
           'dataset' => 'your_dataset_name',
+          'location' => 'asia-northeast1',
           'table' => 'your_table_name',
           'source_format' => 'CSV',
           'max_bad_records' => nil,
@@ -108,6 +109,7 @@ module Embulk
         File.write("tmp/your_file_name", "foobarbaz")
         job_id = Helper.create_load_job_id(task, 'tmp/your_file_name', fields)
         assert job_id.is_a?(String)
+        assert_equal 'embulk_load_job_2abaf528b69987db0224e52bbd1f0eec', job_id
       end
     end
   end

